src/Oscoin/Protocol.hs

```diff
@@ -0,0 +1,276 @@
+import qualified Oscoin.Storage.Block.Orphanage as O
+import           Oscoin.Telemetry (NotableEvent(..))
+import qualified Oscoin.Telemetry as Telemetry
+import           Oscoin.Time.Chrono as Chrono
+
+import           Control.Concurrent.Async (Async)
+import qualified Control.Concurrent.Async as Async
+import           Control.Concurrent.STM
+import qualified Data.List.NonEmpty as NonEmpty
+
+-- | This data structures incorporates all the different components and
+-- data structures in the system devoted to store \"things\". Certain
+-- components (like the block cache, for example) are not managed by the
+-- 'Protocol' directly but they are rather an implementation detail of
+-- whichever 'BlockStore' this manager encapsulates.
+data Protocol c tx s m = Protocol
+    { protoFullBlockStore :: BlockStore c tx s m
+    -- ^ The 'BlockStore', an opaque interface to a component which only
+    -- responsibility is to store and retrieve blocks.
+    , protoOrphanage      :: Orphanage c tx s
+    -- ^ The in-memory data structure which keeps orphans blocks and constructs
+    -- sub-chains suitable for fork selection.
+    , protoValidateFull   :: Validate c tx s
+    -- ^ A validation function to (fully) validate a block.
+    , protoScoreBlock     :: Block c tx (Sealed c s) -> Score
+    -- ^ A function to score blocks.
+    , protoConfig         :: Consensus.Config
+    }
+
+
+-- | A 'Handle' is a record of functions containing two possible
+-- \"strategies\" to dispatch block; calling `dispatchBlockSync` ensures the
+-- function won't return up until the `Protocol` hasn't been fully stepped
+-- (i.e. block inserted in the store, chain selection fully performed etc.)
+-- `dispatchBlockAsync` as the name implies, doesn't block and return the
+-- control to the caller immediately, but offer less integrity guarantees.
+data Handle c tx s m = Handle
+    { dispatchBlockSync  :: Block c tx (Sealed c s) -> m ()
+    -- ^ TODO(Adn): Return an error?
+    , dispatchBlockAsync :: Block c tx (Sealed c s) -> m ()
+    }
+
+{------------------------------------------------------------------------------
+  Using the 'Protocol'
+------------------------------------------------------------------------------}
+
+-- | Given the data necessary to initialise and build the 'Protocol', runs
+-- the input action and return the final result.
+withProtocol
+    :: forall c tx s a m.
+       Orphanage c tx s
+    -> Validate c tx s
+    -> (Block c tx (Sealed c s) -> Score)
+    -> BlockStore c tx s m
+    -> Consensus.Config
+    -> (Protocol c tx s m -> m a)
+    -> m a
+withProtocol o validateFull scoreBlock bs config action =
+    let protocol = Protocol bs o validateFull scoreBlock config
+    in action protocol
+
+-- | Initialises and runs the 'Protocol', by spinning an asynchronous worker
+-- which executes the 'fetchNextBlock' function to pull new blocks from outside
+-- and steps the protocol.
+runOscoinProtocol
```

Comment:
Any reason this function isn't called `runProtocol`, to match `withProtocol`?

Reply:
I just wanted to convey the idea this is like "the thing" we use to run _the_ Oscoin protocol, but for consistency, I'll amend it 😉